### ISSUE TITLE
design: All contour command line arguments should be available in the config file

### ISSUE THREAD

**Issue by stevesloka** [projectcontour/contour#1266]
**Describe the solution you'd like**
#1130 added the ability to have Contour get configuration from a file instead of from only args passed to Contour. 

This issue is to further enhance the support since only two values are plumbed in. Additionally, we should revisit some of the namings to make sure it's clear to users (see #1264). 

Most values are straight forward to wire through, however, some args like `root-ingressroute-namespaces` should take an array.

**Comment by jpeach** [comment]
Related to #2045 and #2044

**Comment by youngnick** [comment]
I've updated the title to ensure that this reflects a broader scope for discussion.

I personally don't have all that strong feelings about this, I think that @jpeach and @davecheney do though. Over to you, gents.

**Comment by youngnick** [comment]
I do feel that the converse is not true: there is definitely some config that does not enter well from the command line (the JSON logging settings being the prime example).

**Comment by davecheney** [comment]
I think there will always be configuration options that make sense in one form and not in another. 

From my point of view uniformity -- having three ways to configure each option, cli flag, config file, env variable -- is not a valuable goal. just writing this made me feel like I had made a mistake adding the configuration file as an option. We should

a. be working hard so that there are usable default values that are so good that they never need over riding
b. only in the cases where we have a documented need should we add a way to override those defaults.

To repeat, the end goal should not be to make Contour maximally configurable. A product that is maximally configurable, is maximally hard to use correctly, offers little opinion and guidance -- just look at the amount of guidance we have to offer with the level of configurability we struggle with with YAML manifests.

In summary, uniformity in configuration options should not be the goal, usability should be the goal. There are some options that make sense in a configuration file because that fits the deployment scenario, there are others that make sense as environment variables if they float naturally out of the environment that the kubelet presents to the pod, and yet more than are easiest expressed as a command line flag. The right answer is contextual, and flows directly from the underlying use case of the feature request.

**Comment by jpeach** [comment]
> I think there will always be configuration options that make sense in one form and not in another.

Yup, there will always be config that only makes sense as a flag (e.g. `--incluster`)

> To repeat, the end goal should not be to make Contour maximally configurable.

That's not what is is being proposed here.

> uniformity in configuration options should not be the goal, usability should be the goal

However, uniformity benefits usability. The close relationship between uniformity and predictability allows users to reason about how to do things and improves their chances of being correct.

> The right answer is contextual.

Yes, of course. Probably the right course of action on this issue is to review the configuration and hoist elements up to the configuration file where there is not a good argument for leaving it as command-line only.

**Comment by sunjayBhatia** [comment]
@projectcontour/maintainers can this be closed? since we're moving to a future where we have minimal CLI args and a dynamic config CRD for other arguments

**Comment by stevesloka** [comment]
Agree, I think we can close.
